src/gc.rs

```diff
@@ -0,0 +1,92 @@
+use crate::Sodg;
+use anyhow::{Context, Result};
+use std::collections::VecDeque;
+
+impl Sodg {
+    /// Attempts to collect the vertex `v`.
+    ///
+    /// If there are no edges leading to `v`, then it is deleted and all its children are collected.
+    /// Otherwise, nothing happens.
+    ///
+    /// ```
+    /// use sodg::{Hex, Sodg};
+    /// let mut g = Sodg::empty();
+    /// g.add(1).unwrap();
+    /// g.put(1, Hex::from(0)).unwrap();
+    /// g.add(2).unwrap();
+    /// g.put(2, Hex::from(0)).unwrap();
+    /// g.bind(1, 2, "x").unwrap();
+    /// g.data(2).unwrap(); // Try to collect 2
+    /// assert!(g.data(2).is_ok());
+    /// g.data(1).unwrap(); // Successfully collect 1
+    /// assert!(g.data(1).is_err());
+    /// ```
+    pub(crate) fn collect(&mut self, v: u32) -> Result<()> {
+        let mut queue = VecDeque::new();
+        queue.push_back(v);
+        while !queue.is_empty() {
+            let collected_vertex_id = queue
+                .pop_front()
+                .context("A non-empty queue failed to yield an element, this shouldn't happen.")?;
+            let collected_vertex = self
+                .vertices
+                .get(&collected_vertex_id)
+                .context(format!("Failed to get v{collected_vertex_id}"))?
+                .clone();
+            if collected_vertex.parents.is_empty() {
+                for edge in &collected_vertex.edges {
+                    queue.push_back(edge.to);
+                    self.vertices
+                        .get_mut(&edge.to)
+                        .context(format!("Failed to get v{}", edge.to))?
+                        .parents
+                        .remove(&collected_vertex_id);
+                }
+                self.vertices.remove(&collected_vertex_id);
```

Comment:
@UARTman there seems to be a mistake in the algorithm. What if we have a vertex that is supposed to be collected, but it has kids that are not supposed to be collected: their data has not been read yet. The current algorithm will delete them all.